lib/util/lib/platform.c

```diff
@@ -197,10 +198,48 @@ hse_cpu_init(void)
     return 0;
 }
 
+unsigned long
+platform_lowmem_scale(const unsigned long memgb)
+{
+    return max_t(unsigned long, 1, roundup_pow_of_two(memgb) / HSE_LOWMEM_THRESHOLD_GB_MIN);
+}
+
+static void
+hse_lowmem_adjust(unsigned long *memgb)
+{
+    struct hse_gparams gpdef = hse_gparams_defaults();
+    unsigned long      mavail = 0;
+
+    /* [HSE_REVISIT] mapi breaks initialization of mavail.
+     */
+    hse_meminfo(NULL, &mavail, 30);
+
+    if (mavail <= HSE_LOWMEM_THRESHOLD_GB_DFLT) {
+        unsigned long scale = platform_lowmem_scale(mavail);
+
+        /* Scale various caches based on the available memory */
+        if (hse_gparams.gp_c0kvs_ccache_sz_max == gpdef.gp_c0kvs_ccache_sz_max)
+            hse_gparams.gp_c0kvs_ccache_sz_max =
+                min_t(uint64_t, HSE_C0_CCACHE_SZ_MIN * scale, HSE_C0_CCACHE_SZ_MAX);
+
+        if (hse_gparams.gp_c0kvs_cheap_sz == gpdef.gp_c0kvs_cheap_sz)
+            hse_gparams.gp_c0kvs_cheap_sz =
+                min_t(uint64_t, HSE_C0_CHEAP_SZ_MIN * scale, HSE_C0_CHEAP_SZ_MAX);
+
+        if (hse_gparams.gp_vlb_cache_sz == gpdef.gp_vlb_cache_sz)
+            hse_gparams.gp_vlb_cache_sz =
+                min_t(uint64_t, HSE_VLB_CACHESZ_MIN * scale, HSE_VLB_CACHESZ_MAX);
+    }
+
+    assert(memgb);
+    *memgb = mavail;
+}
```

Comment:
These really aren't platform specific...

Reply:
You can't use the lowmem functions before initializing event counters and the data tree. All suggestions are welcome.